### ISSUE TITLE
Filterable behavior for Image Block raw transformations

### ISSUE THREAD

**Issue by danielbachhuber** [WordPress/gutenberg#8473]
One of the plugins I work on, Tasty Pins, offers the ability to add "Pinterest Text" to an image (related #8472). When the image ends up in the Classic Editor, its markup looks like this:

<img width="791" alt="image" src="https://user-images.githubusercontent.com/36432/43667825-8f5e4522-972e-11e8-9082-33f6c94e246e.png">

Notice the inclusion of `data-pin-description="This cashew coffee is delicious, creamy, and easy. Perfect for a chilly morning! | pinchof yum.com"`.

When this image is transformed to an Image Block, the `data-pin-description` attribute is lost, even though I've registered it as a custom attribute to the block (related #8470).

![pinterestconversion](https://user-images.githubusercontent.com/36432/43668851-024686d0-9734-11e8-9f94-29fc41746a32.gif)

What's the best way to filter the Image Block transform behavior? It doesn't appear plugins have much control over the behavior:

https://github.com/WordPress/gutenberg/blob/804abecb7fee1acfbc764a9d6a1028ecf55b72c3/core-blocks/image/index.js#L112-L125

**Comment by ngygen323** [comment]
Read the reviews of people who are against it. Usually a project gets hate for the change. But in this case I feel that everyone has a fair point for the opposition. Text editor is a vital part and it should not be like this. Is this purely driven by core developers? I see you are ignoring a lot of responses.Read the reviews of people who are against it. Usually a project gets hate for the change. But in this case I feel that everyone has a fair point for the opposition. Text editor is a vital part and it should not be like this. Is this purely driven by core developers? I see you are ignoring a lot of responses.Read the reviews of people who are against it. Usually a project gets hate for the change. But in this case I feel that everyone has a fair point for the opposition. Text editor is a vital part and it should not be like this. Is this purely driven by core developers? I see you are ignoring a lot of responses.Read the reviews of people who are against it. Usually a project gets hate for the change. But in this case I feel that everyone has a fair point for the opposition. Text editor is a vital part and it should not be like this. Is this purely driven by core developers? I see you are ignoring a lot of responses.Read the reviews of people who are against it. Usually a project gets hate for the change. But in this case I feel that everyone has a fair point for the opposition. Text editor is a vital part and it should not be like this. Is this purely driven by core developers? I see you are ignoring a lot of responses.Read the reviews of people who are against it. Usually a project gets hate for the change. But in this case I feel that everyone has a fair point for the opposition. Text editor is a vital part and it should not be like this. Is this purely driven by core developers? I see you are ignoring a lot of responses.Read the reviews of people who are against it. Usually a project gets hate for the change. But in this case I feel that everyone has a fair point for the opposition. Text editor is a vital part and it should not be like this. Is this purely driven by core developers? I see you are ignoring a lot of responses.Read the reviews of people who are against it. Usually a project gets hate for the change. But in this case I feel that everyone has a fair point for the opposition. Text editor is a vital part and it should not be like this. Is this purely driven by core developers? I see you are ignoring a lot of responses.Read the reviews of people who are against it. Usually a project gets hate for the change. But in this case I feel that everyone has a fair point for the opposition. Text editor is a vital part and it should not be like this. Is this purely driven by core developers? I see you are ignoring a lot of responses.

**Comment by designsimply** [comment]
Noting I deleted a comment which was saying the same thing over and over many times per minute in various locations. Spamming and unacceptable behavior towards anyone, or this community, will not be allowed. If you want to give feedback, please do it without spamming or personal attacks.

My apologies for those who were effected during this. Please know we have a zero tolerance policy over this type of behavior.

**Comment by mcsf** [comment]
> HTML is unexpectedly sanitized before the transformation is called

@danielbachhuber, is this helpful?

https://github.com/WordPress/gutenberg/issues/8648#issuecomment-434781445

**Comment by danielbachhuber** [comment]
> @danielbachhuber, is this helpful?
> 
> [#8648 (comment)](https://github.com/WordPress/gutenberg/issues/8648#issuecomment-434781445)

Yes, I think this would solve the problem I ran into with https://github.com/WordPress/gutenberg/issues/8473#issuecomment-433135779. However, I want to see how #11440 transforms into actionable tasks before I spend too much time on this.

**Comment by danielbachhuber** [comment]
Along with https://github.com/WordPress/gutenberg/issues/10204#issuecomment-439254269, I've gone with this:

```
const {
	merge
} = lodash;

const { hooks } = wp;

const modifyTransforms = ( settings, name ) => {

	if ( 'core/image' !== name ) {
		return settings;
	}

	const imageAttributes = [
		'src',
		'alt',
		'data-pin-description',
	];

	settings.transforms.from[0] = merge( settings.transforms.from[0], {
		schema: {
			figure: {
				children: {
					a: {
						children: {
							img: {
								attributes: imageAttributes
							}
						}
					},
					img: {
						attributes: imageAttributes
					}
				}
			}
		}
	} );

	return settings;
};

hooks.addFilter( 'blocks.registerBlockType', 'wp-tasty/tasty-pins', modifyTransforms );
```

It doesn't work for images with captions though.
